File: src/main/java/com/machfour/macros/cli/modes/Recipe.java
package com.machfour.macros.cli.modes;

import com.machfour.macros.cli.CommandImpl;
import com.machfour.macros.cli.utils.CliUtils;
import com.machfour.macros.ingredients.IngredientsParser;
import com.machfour.macros.linux.Config;
import com.machfour.macros.linux.LinuxDatabase;
import com.machfour.macros.objects.CompositeFood;
import com.machfour.macros.objects.NutritionData;
import com.machfour.macros.storage.MacrosDatabase;

import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import static com.machfour.macros.linux.Config.PROGNAME;

public class Recipe extends CommandImpl {
    private static final String NAME = "recipe";
    private static final String USAGE = String.format("Usage: %s %s <recipes.json>", PROGNAME, NAME);

    public Recipe() {
        super(NAME, USAGE);
    }

    @Override
    public void doActionNoExitCode(List<String> args) {
        if (args.contains("--help")) {
            printHelp();
            return;
        } else if (args.size() < 2) {
            out.println(usage());
            return;
        }

        MacrosDatabase db = LinuxDatabase.getInstance(Config.DB_LOCATION);
        List<CompositeFood> recipes = new ArrayList<>();

        try (Reader jsonReader = new FileReader(args.get(1))) {
            out.println("Importing recipes...");
            recipes.addAll(IngredientsParser.readRecipes(jsonReader, db));
        } catch (IOException e1) {
            out.println("IO exception occurred while reading recipes file: " + e1.getLocalizedMessage());
        } catch (SQLException e2) {
            out.println("SQL exception occurred while creating recipe objects: " + e2.getMessage());
        }

        if (recipes.isEmpty()) {
            out.println("No recipes read! Check the recipes file");
            return;
        }

        out.println("The following recipes were found:");

        for (CompositeFood cf : recipes) {
            out.println();
            out.println("Name: " + cf.getMediumName());
            out.println();
            out.println("Ingredients:");
            out.println();
            CliUtils.printIngredients(cf.getIngredients(), out);
            out.println();
            out.println("Nutrition Information:");
            out.println();
            NutritionData nd = cf.getNutritionData();
            String unit = nd.qtyUnitAbbr();
            // if entered not per 100g, print both original amount and per 100 g
            if (nd.getQuantity() != 100) {
                out.printf("Per %.0f%s:\n", nd.getQuantity(), unit);
                CliUtils.printNutritionData(nd, false, out);
                out.println();
                nd = nd.rescale(100);
            }
            out.printf("Per %.0f%s:\n", nd.getQuantity(), unit); // should now be 100
            CliUtils.printNutritionData(nd, false, out);
            out.println();
            out.println("================================================");
            out.println();
        }

        String article = recipes.size() == 1 ? "this" : "these";
        String plural = recipes.size() == 1 ? "" : "s";
        out.print("Would you like to save " + article + " food" + plural + "? [y/N] ");
        char response = CliUtils.getChar(in, out);
        out.println();
        if (response == 'y' || response == 'Y') {
            try {
                IngredientsParser.saveRecipes(recipes, db);
                out.println("Recipes saved!");
            } catch (SQLException e) {
                out.println("SQL exception occurred while saving recipe objects: " + e.getMessage());
                out.println("Recipes not saved");
            }
        } else {
            out.println("Okay.");
        }
    }

}

File: src/main/java/com/machfour/macros/cli/modes/Recipe.kt
package com.machfour.macros.cli.modes

import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.cliGetChar
import com.machfour.macros.cli.utils.printIngredients
import com.machfour.macros.cli.utils.printNutrientData
import com.machfour.macros.core.MacrosConfig
import com.machfour.macros.entities.CompositeFood
import com.machfour.macros.ingredients.IngredientsParser
import java.io.FileReader
import java.io.IOException
import java.sql.SQLException


class Recipe(config: MacrosConfig): CommandImpl(NAME, USAGE, config) {
    companion object {
        private const val NAME = "recipe"
        private const val USAGE = "Usage: $programName $NAME <recipes.json>"
    }

    override fun doAction(args: List<String>) : Int {
        if (args.contains("--help")) {
            printHelp()
            return -1
        } else if (args.size < 2) {
            println(usage)
            return -1
        }

        val ds = config.database
        val recipes = ArrayList<CompositeFood>()

        try {
            FileReader(args[1]).use { jsonReader ->
                println("Importing recipes...")
                recipes.addAll(IngredientsParser.readRecipes(jsonReader, ds))
            }
        } catch (e1: IOException) {
            println("IO exception occurred while reading recipes file: " + e1.message)
            return 1
        } catch (e2: SQLException) {
            println("SQL exception occurred while creating recipe objects: " + e2.message)
            return 1
        }

        if (recipes.isEmpty()) {
            println("No recipes read! Check the recipes file")
            return 2
        }

        println("The following recipes were found:")

        for (cf in recipes) {
            println()
            println("Name: " + cf.mediumName)
            println()
            println("Ingredients:")
            println()
            printIngredients(cf.ingredients)
            println()
            println("Nutrition Information:")
            println()
            val nd = cf.nutrientData
            val unit = nd.qtyUnitAbbr
            // if entered not per 100g, print both original amount and per 100 g
            if (nd.quantity != 100.0) {
                println("Per %.0f%s:".format(nd.quantity, unit))
                printNutrientData(nd, false)
                println()
            }
            println("Per %.0f%s:".format(nd.quantity, unit)) // should now be 100
            printNutrientData(nd.rescale100(), false)
            println()
            println("================================================")
            println()
        }

        val article = if (recipes.size == 1) "this" else "these"
        val plural = if (recipes.size == 1) "" else "s"
        print("Would you like to save $article food$plural? [y/N] ")
        val response = cliGetChar()
        println()
        if (response == 'y' || response == 'Y') {
            try {
                IngredientsParser.saveRecipes(recipes, ds)
                println("Recipes saved!")
            } catch (e: SQLException) {
                println("SQL exception occurred while saving recipe objects: " + e.message)
                println("Recipes not saved")
                return 1
            }
        } else {
            println("Okay.")
        }
        return 0
    }
}
